### ISSUE TITLE
Request for improved SSL server certificate verification

### ISSUE THREAD

**Issue by sodabrew** [eventmachine/eventmachine#814]
From https://github.com/faye/faye-websocket-ruby/pull/101#issuecomment-349894343

I'm not entirely sure if this is the right PR/issue to discuss this on, but I can briefly summarize what EM support I think would be required for implementing SSL clients with server certificate verification,

Listed in terms of the exposed libssl APIs:

 * [SSL_CTX_load_verify_locations](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_load_verify_locations.html) Supported in eventmachine/eventmachine#378 by extending EventMachine::set_tls_parms

* [SSL_CTX_set_verify](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_verify.html) with SSL_VERIFY_PEER: already supported in EM: https://github.com/eventmachine/eventmachine/blob/bc9e2faefafb7c13c3b4baeacb9763878a26a348/ext/ssl.cpp#L341-L346

* The `SSL_set_verify(..., ssl_verify_wrapper)` callback MUST return false by default if called with `!preverify_ok`.

The `ssl_verify_wrapper` ignoring the `preverify_ok` parameter is the most blatantly broken part of the implementation, because this effectively bypasses all of the libssl certificate validation logic 👿

Based on my reading of the docs and issues like eventmachine/eventmachine#275, I suspect this even includes very fundamental things like "the private key used to sign the session key matches the public key in the certificate".

* [SSL_get_peer_certificate](https://www.openssl.org/docs/manmaster/man3/SSL_get_peer_certificate.html): already supported in https://github.com/eventmachine/eventmachine/blob/bc9e2faefafb7c13c3b4baeacb9763878a26a348/ext/ssl.cpp#L532-L540 ?

I think the` SSL_CTX_load_verify_locations` + `preverify_ok` changes would be the bare minimum that would be required. These also match the changes dicussed/implemented in eventmachine/eventmachine#378

Additional bonus points for:

SSL_get_verify_result + X509_verify_cert_error_string to allow the application to report more useful error messages than just "certificate verification failed"

Some convenience wrapper for the cert subject/hostname validation - ideally there should be some kind of secure: true/false boolean that doesn't require each client developer to research and write their own certificate verification wrappers for the vast majority of usecases

I think Ruby's [OpenSSL::SSL.verify_certificate_identity](https://github.com/ruby/openssl/blob/308c158d7be4208591519a4d276014178e84f434/lib/openssl/ssl.rb#L262-L344) can probably be used by applications together with the `SSL_get_peer_certificate` API, so it doesn't necessarily need to be part of EM itself.

**Comment by parhs** [comment]
I am looking for this functionality for hours but is there a good approach with current version to verify somehow safely connections?

**Comment by sodabrew** [comment]
No, sorry. The current versions do not offer complete cert verification.

**Comment by parhs** [comment]
@sodabrew Are there any changes that can be done easily to increase security ?
